File: PE_0407/PE_0407.py
# -*- coding: utf-8 -*-
"""

PE_0407

Idempotents

Created on Tue Jan 10 04:49:44 2017
@author: mbh
"""

import itertools as it
import time
import numba as nb
import numpy as np
from numba.typed import List

def main(limit=10**7):
    t0=time.perf_counter()
    p407(limit)
    print(time.perf_counter()-t0)

# @nb.njit    
def p407(limit):
    misum=0
    for n in range(2,limit+1):
        nsum=max_idempotent(n)
        misum+=nsum
    print(misum)

# @nb.njit  
def max_idempotent(n):
    """returns maximum idempotent a < n: a^2=a mod n"""    
    pfs=pflist(n)
    pfnum=len(pfs)
    if pfnum==1:
        return 1 #idempotent=1 for primes or powers of primes
        
    #Use the CRT to find m 'base' idempotent solutions from m prime factors p_i^a_i   
    idems=[]
    for i in range(pfnum):
        allButOnePfs=pfs[:i]+pfs[i+1:]
        xsum=0
        for i in range(pfnum-1):
            Ni=n//allButOnePfs[i]
            xsum+=inverse(Ni,allButOnePfs[i])*Ni        
        idems.append(xsum % n)

    #generate all other idempotents from these, and return the maximum
    maxval=max(idems)
    for i in range(2,len(idems)):
        for a in it.combinations(idems, i):
        # combs=comb2([],idems, i,[])
        # print(combs)
        # for a in combs:
            aprod=1
            for x in a:
                aprod = (aprod * x) % n
            if aprod>maxval:
                maxval=aprod
    # return maxval
    return np.int64(maxval)

@nb.njit   
def pflist(n):
    """returns the distinct prime factors of n as [2^a,3^b.....]"""   
    i = 2
    factors = []
    while i * i <= n:
        if n % i:
            i += 1
        else:
            factors.append(1)
            while not n %i:
                n //= i
                factors[-1]*=i
    if n > 1:
        factors.append(n)
    return factors            

@nb.njit    
def inverse(a, n):
    """returns multiplicative inverse of a mod n. a and n must be-co-prime"""
    t1,t2=0,1    
    r1,r2=n,a    
    while r2!=0:
        q = r1 // r2
        t1, t2 = t2, t1 - q * t2
        r1, r2 = r2, r1 - q * r2
    if t1 < 0:
        t1 +=n
    return t1 
    
#not used
def maxIdem(s,n):
    maxval=max(s)
    for i in range(2,len(s)):
        for a in it.combinations(s, i):
            aprod=1
            for x in a:
                aprod*=x
                aprod=aprod%n
            if aprod>maxval:
                maxval=aprod
    return maxval




def comb(sofar, rest, n):
    if n == 0:
        print(sofar)
    else:
        for i in range(len(rest)):
            comb(sofar + rest[i], rest[i+1:], n-1)

# @nb.njit
def comb2(sofar, rest, n,results):
    if n == 0:
        results.append(sofar)
    else:
        for i in range(len(rest)):
            comb2(sofar+[rest[i]], rest[i+1:], n-1,results)
    return results

            
def comb3(list_of_values,n):
    sofar=[]
    
    while n>=0:
        pass
        



            
def test():
    
    for a in comb2([],[1,2,3],2):
        print(a)

File: PE_0407/PE_0407.jl
# -*- coding: utf-8 -*-
"""

PE_0407

Idempotents

Created on Tue Jan 10 04:49:44 2017
@author: mbh
"""


using Mods
using Primes
using Combinatorics

function test(idem=[1,2,3])
    # for a in Iterators.product(fill(idem,2)...)
    for a in combinations(idem,2)
        println(a)
    end
end

function main(limit=10^7)
    a=p407(10)
    @time p407(limit)
end


function p407(limit)
    misum=0
    for n in 2:limit
        nsum=max_idempotent(n)
        misum+=nsum
    end
    return(misum)
end


function max_idempotent(n)
    #returns maximum idempotent a < n: a^2=a mod n"""
    pfs=pflist(n)
    pfnum=length(pfs)
    if pfnum==1
        return 1 #idempotent=1 for primes or powers of primes
    end

    #Use the CRT to find m 'base' idempotent solutions from m prime factors p_i^a_i
    idems=Int64[]
    for i in 1:pfnum
        allButOnePfs=vcat(pfs[1:i-1],pfs[i+1:end])
        xsum=0
        for i in 1:pfnum-1
            Ni=div(n,allButOnePfs[i])
            xsum+=inverse(Ni,allButOnePfs[i])*Ni # faster
        end
        push!(idems,xsum % n)
    end
    #generate all other idempotents from these, and return the maximum
    maxval=maximum(idems)
    for i in 2:length(idems)-1
        for a in combinations(idems,i)
            aprod=1
            for x in a
                aprod = (aprod * x) % n
            end
            if aprod>maxval
                maxval=aprod
            end
        end
    end
    return maxval
end

#from Wikipedia
function inverse(a, n)
    #returns multiplicative inverse of a mod n. a and n must be-co-prime
    t1,t2=0,1
    r1,r2=n,a
    while r2!=0
        q = div(r1,r2)
        t1, t2 = t2, t1 - q * t2
        r1, r2 = r2, r1 - q * r2
    end
    if t1 < 0
        t1 +=n
    end
    return t1
end

function pflist(n)
    # returns the distinct prime factors of n as [2^a,3^b.....]
    i = 2
    factors = Int64[]
    while i * i <= n
        if n % i>0
            i += 1
        else
            push!(factors,1)
            while n %i ==0
                n =div(n,i)
                factors[end]*=i
            end
        end
    end
    if n > 1
        push!(factors,n)
    end
    return factors
end
